src/active_item_cache.cpp

```diff
@@ -11,26 +12,45 @@ void active_item_cache::remove( std::list<item>::iterator it, point location )
     // But the processing_speed may have changed, so if it's not in the expected container,
     // remove it from all containers to ensure no stale iterator remains.
     auto &expected = active_items[it->processing_speed()];
-    const auto iter = std::find_if( expected.begin(), expected.end(), predicate );
+    auto iter = std::find_if( expected.begin(), expected.end(), predicate );
     if( iter != expected.end() ) {
+        active_item_set.erase( iter->item_id );
         expected.erase( iter );
-    } else {
-        for( auto &e : active_items ) {
-            e.second.remove_if( predicate );
+        return;
+    }
+    for( auto &e : active_items ) {
+        iter = std::find_if( e.second.begin(), e.second.end(), predicate );
+        if( iter != e.second.end() ) {
+            active_item_set.erase( iter->item_id );
+            e.second.erase( iter );
+            return;
         }
     }
-    active_item_set.erase( &*it );
+    debugmsg( "Critical: The item isn't there!" );
 }
 
 void active_item_cache::add( std::list<item>::iterator it, point location )
 {
-    active_items[it->processing_speed()].push_back( item_reference{ location, it, &*it } );
-    active_item_set.insert( &*it );
+    // Uniqueness is guaranteed because active item processing always removes the item,
+    // processes it, then sometimes add it back.
+    assert( active_item_set.count( next_free_id ) == 0 );
+
+    active_items[it->processing_speed()].push_back( item_reference { location, it, next_free_id } );
+    active_item_set.insert( next_free_id );
+    // Blindly incrementing is safe because the C++ standard says that unsigned ints wrap around.
+    ++next_free_id;
 }
 
-bool active_item_cache::has( std::list<item>::iterator it, point ) const
+bool active_item_cache::has( std::list<item>::iterator it, point p ) const
 {
-    return active_item_set.count( &*it ) != 0;
+    const auto predicate = [&]( const item_reference & ir ) {
+        // Comparing iterators from different sequences is undefined, so check the point first
+        return ir.location == p && it == ir.item_iterator;
+    };
+    return std::any_of( active_items.begin(), active_items.end(),
+    [&]( const std::pair< int, std::list<item_reference> > &elem ) {
+        return std::any_of( elem.second.begin(), elem.second.end(), predicate );
+    } );
```

Comment:
Thanks for the review, I'd be glad to clean it up some more.

Something like this?
```
/**
 * Search active_items[key] for a certain element
 * Returns active_items[key].end() if not found
*/
std::list<item_reference>::iterator find( int key, const std::list<item>::iterator &it,
                                          const point &p) const;
```